onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -36,6 +37,163 @@ CoreMLExecutionProvider::CoreMLExecutionProvider(const CoreMLOptions& options)
+    // Get node arg type
+    const auto* node_arg = graph_viewer.GetNodeArg(io_name);
+    int32_t type = ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
+    if (!GetType(*node_arg, type, logger)) {
+      LOGS(logger, ERROR) << "NodeArg " << io_name << " has no type information.";
+      continue;
+    }
+
+    // Get relevant nodes (producers or consumers)
+    std::vector<const Node*> nodes;
+    if (is_input) {
+      nodes = graph_viewer.GetConsumerNodes(io_name);
+    } else {
+      const Node* producer = graph_viewer.GetProducerNode(io_name);
+      if (producer) nodes.push_back(producer);
+    }
+
+    // Process nodes based on type compatibility
+    if (!IsSupportedInputOrOutputType(type)) {
+      for (const Node* node : nodes) {
+        if (node && partition_nodes.find(node->Index()) != partition_nodes.end()) {
+          update_meta_def = true;
+          partition_nodes.erase(node->Index());
+
+          // Update the appropriate set after node removal
+          if (is_input) {
+            UpdateInputsSetAfterNodeRemoval(partition_nodes, node, new_names, graph_viewer);
+          } else {
+            UpdateOutputsSetAfterNodeRemoval(partition_nodes, node, new_names, graph_viewer);
+          }
+
+          LOGS(logger, VERBOSE) << "Removing node " << node->Name()
+                                << " from CoreML partition due to unsupported type: "
+                                << ONNX_NAMESPACE::TensorProto_DataType_Name(
+                                       (ONNX_NAMESPACE::TensorProto_DataType)type);
+        }
+      }
+    } else {
+      // Type is supported, check if nodes are in partition
+      for (const Node* node : nodes) {
+        if (node && partition_nodes.find(node->Index()) != partition_nodes.end()) {
+          new_names.insert(io_name);
+        } else {
+          update_meta_def = true;
+        }
+      }
+    }
+  }
+
+  // Update metadata if needed
+  if (update_meta_def) {
+    io_names = std::vector<std::string>(new_names.begin(), new_names.end());
+  }
+
+  return original_partition_size != partition_nodes.size();
+}
+
+void CoreMLExecutionProvider::FilterIncompatibleEdgeNodesFromPartition(IndexedSubGraph& partition,
+                                                                       const onnxruntime::GraphViewer& graph_viewer,
+                                                                       const logging::Logger& logger) const {
+  IndexedSubGraph::MetaDef* meta_def = partition.GetMutableMetaDef();
+  if (!meta_def) return;
+
+  std::unordered_set<NodeIndex> partition_nodes(partition.nodes.begin(), partition.nodes.end());
+  bool modified;
+
+  do {
+    bool inputs_modified = ProcessIncompatibleNodes(graph_viewer, partition_nodes, meta_def, true, logger);
+    bool outputs_modified = ProcessIncompatibleNodes(graph_viewer, partition_nodes, meta_def, false, logger);
+
+    modified = inputs_modified || outputs_modified;
+  } while (modified);
```

Comment:
it seems like this could be inefficient. e.g., if input processing and output processing don't modify the partition (i.e., return true) the same number of times, then some iterations will be unnecessary for the one that finishes earlier.